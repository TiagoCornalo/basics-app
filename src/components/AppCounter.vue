<template>
  <div>
    <p data-testid="counter">{{ count }}</p>
    <button @click="count++">Increment</button>
    <button @click="count--">Decrement</button>
    <button @click.alt="count += 10">Increment 10</button>
  </div>
  <div>
    <h2>
      {{ customTitle }}
    </h2>
    <p> {{ count }} <sup>2</sup> = {{ squareValue }}</p>
    <p>{{ count }}</p>
  </div>
</template>

<script>
export default {
  props: {
    title: String,
    start: {
      type: Number,
      default: 150
    }
  },
  name: 'AppCounter',

  data() {
    return {
      count: this.start
    }
  },

  methods: {
    getSquareValue() {
      return this.count * this.count;
    }
  },

  computed: {
    squareValue() {
      return this.count * this.count;
    },
    customTitle() {
      return this.title || 'Counter';
    }
  }
}
</script>

<style></style>