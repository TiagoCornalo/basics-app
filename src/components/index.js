export { default as AppCounter } from './AppCounter.vue';
export { default as AppIndecision } from './AppIndecision.vue';
