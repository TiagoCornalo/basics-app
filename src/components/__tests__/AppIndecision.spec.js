import { describe, it, expect, beforeEach, vi, afterEach } from 'vitest'
import { shallowMount } from '@vue/test-utils'
import { AppIndecision } from '@/components'

describe('AppIndecision', () => {
  let wrapper;
  let getAnswerMock;

  vi.spyOn(window, 'fetch').mockImplementation(() => {
    return Promise.resolve({
      json: () => Promise.resolve({
        "answer": "yes",
        "forced": false,
        "image": "https://yesno.wtf/assets/yes/2.gif"
      })
    })
  })

  beforeEach(() => {
    wrapper = shallowMount(AppIndecision)

    getAnswerMock = vi.spyOn(wrapper.vm, 'getAnswer')
  })

  afterEach(() => {
    getAnswerMock.mockRestore()
  })


  it('should match snapshot', () => {
    expect(wrapper.html()).toMatchSnapshot()
  })

  it('should not fire http request when writing', () => {
    const input = wrapper.find('input')
    input.setValue('Hello')

    expect(getAnswerMock).not.toHaveBeenCalled()
  })

  it('should fire http request when writing "?" symbol', async () => {
    const input = wrapper.find('input')
    await input.setValue('Hello?')

    expect(getAnswerMock).toHaveBeenCalled(1)
  })

  describe('getAnswer', () => {

    it('should return an answer and an image', async () => {
      await wrapper.vm.getAnswer()

      expect(wrapper.vm.answer).toBe('Sí!')
      expect(wrapper.vm.img).toBe('https://yesno.wtf/assets/yes/2.gif')
    })

    it('api fails', async () => {
      fetch.mockImplementationOnce(() => Promise.reject('API is down'))

      await wrapper.vm.getAnswer()

      expect(wrapper.vm.answer).toBe('Ups! Algo salió mal')
      expect(wrapper.vm.img).toBeFalsy()
    })
  })
})