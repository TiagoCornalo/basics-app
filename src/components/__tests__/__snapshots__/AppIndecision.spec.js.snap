// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`AppIndecision > should match snapshot 1`] = `
"<!--v-if-->
<div data-v-40bba867="" class="bg-dark"></div>
<div data-v-40bba867="" class="indecision-container"><input data-v-40bba867="" type="text" placeholder="Hazme una pregunta">
  <p data-v-40bba867="">Recuerda terminar con un signo de interrogaci√≥n (?)</p>
  <!--v-if-->
</div>"
`;
